tests/AsyncTest.php

```diff
@@ -84,4 +85,65 @@ public function testAsyncReturnsPromiseThatFulfillsWithValueWhenCallbackReturnsA
         $this->assertGreaterThan(0.1, $time);
         $this->assertLessThan(0.12, $time);
     }
+
+    public function testCancel()
+    {
+        self::expectOutputString('a');
+
+        $promise = async(function (): int {
+            echo 'a';
+            await(sleep(2));
+            echo 'b';
+
+            return time();
+        })();
+        Loop::addTimer(0.5, function () use ($promise) {
+            $promise->cancel();
+        });
+
+        Loop::run();
+    }
+
+    public function testNestedCancel()
+    {
+        self::expectOutputString('abc');
+
+        $promise = async(function (): int {
+            echo 'a';
+            await(async(function(): void {
+                echo 'b';
+                await(async(function(): void {
+                    echo 'c';
+                    await(sleep(2));
+                    echo 'd';
+                })());
+                echo 'e';
+            })());
+            echo 'f';
+
+            return time();
+        })();
+        Loop::addTimer(0.5, function () use ($promise) {
+            $promise->cancel();
+        });
+
+        Loop::run();
+    }
```

Comment:
@azjezz Just added a test for your https://github.com/reactphp/async/pull/20#issuecomment-1007258134